gibbon-compiler/src/Packed/FirstOrder/L3/Syntax.hs

```diff
@@ -44,12 +44,16 @@ data E3Ext loc dec =
   | AddCursor Var (L (E3 loc dec)) -- ^ Add a constant offset to a cursor variable
   | ReadTag   Var                  -- ^ One cursor in, (tag,cursor) out
   | WriteTag  DataCon Var          -- ^ Write Tag at Cursor, and return a cursor
-  | NewBuffer                      -- ^ Create a new buffer, and return a cursor
-  | ScopedBuffer                   -- ^ Create a temporary scoped buffer, and
+  | NewBuffer Multiplicity         -- ^ Create a new buffer, and return a cursor
+  | ScopedBuffer Multiplicity      -- ^ Create a temporary scoped buffer, and
                                    --   return a cursor
+  | InitSizeOfBuffer Multiplicity  -- ^ Returns the initial buffer size for a specific multiplicity
+                                   --
   | SizeOfPacked Var Var           -- ^ Takes in start and end cursors, and returns an Int
                                    --   we'll probably represent (sizeof x) as (end_x - start_x) / INT
   | SizeOfScalar Var               -- ^ sizeof(var)
+  | BoundsCheck TyCon Int Var Var  -- ^ Tycon, size of scalars, region, cursor
+  | ReadCursor Var                 -- ^ Reads and returns the cursor at Var
```

Comment:
`BoundsCheck` is the same as L2. `ReadCursor` is used to read redirection pointers.